lib/FrameManager.js

```diff
@@ -810,7 +810,11 @@ class WaitTask {
     let success = null;
     let error = null;
     try {
-      success = await (await this._frame.executionContext()).evaluateHandle(waitForPredicatePageFunction, this._predicateBody, this._polling, this._timeout, ...this._args);
+      // Do surgery over evaluation string to insert predicate body and make WaitTask work with strict CSP.
+      let evalFunctionText = waitForPredicatePageFunction.toString();
+      evalFunctionText = evalFunctionText.replace('// _PUPPETEER_PREDICATE_BODY_', this._predicateBody);
+      const evalFunction = eval('var x = ' + evalFunctionText + ';x');
```

Comment:
just to make things a bit more straightforward:
```js
const evalFunction = eval('(' + evalFunctionText + ')');
```

Reply:
Something like this I think will fix the scoped variables from waitForPredicateFunction polluting the user code.
```js
function thing() {
(**waitForPredicateFunction**).call(null, function(..args){**predicate_body**}, ...arguments)
}```